File: agent-and-tools/apps/prompt-composer/src/lib/audit-gov-emit.ts
/**
 * M22 / fan-out — fire-and-forget emitter to audit-governance-service.
 *
 * Producers should NEVER await this — emission failures must not block any
 * request handler. Set AUDIT_GOV_URL="" to disable.
 */
import { logger } from "../config/logger";

const AUDIT_GOV_URL = process.env.AUDIT_GOV_URL ?? "http://host.docker.internal:8500";
const TIMEOUT_MS    = 5_000;
const AUDIT_GOV_SERVICE_TOKEN = process.env.AUDIT_GOV_SERVICE_TOKEN ?? "";

function auditHeaders(): Record<string, string> {
  return AUDIT_GOV_SERVICE_TOKEN
    ? { "content-type": "application/json", authorization: `Bearer ${AUDIT_GOV_SERVICE_TOKEN}` }
    : { "content-type": "application/json" };
}

/**
 * M35.4 — trace_id is now mandatory (TypeScript-level required field).
 * Pass `undefined` only when a trace genuinely cannot be derived. We log
 * a warning at runtime so call sites that lose their trace_id surface.
 */
export interface EmitInput {
  trace_id:       string | undefined;
  source_service: string;
  kind:           string;
  subject_type?:  string;
  subject_id?:    string;
  actor_id?:      string;
  capability_id?: string;
  tenant_id?:     string;
  severity?:      "info" | "warn" | "error" | "audit";
  payload?:       Record<string, unknown>;
}

export function emitAuditEvent(input: EmitInput): void {
  if (!AUDIT_GOV_URL) return;
  if (!input.trace_id) {
    logger.warn(`audit-gov emit ${input.source_service}/${input.kind} missing trace_id — event will not be joinable to a run`);
  }
  void (async () => {
    try {
      const res = await fetch(`${AUDIT_GOV_URL.replace(/\/$/, "")}/api/v1/events`, {
        method:  "POST",
        headers: auditHeaders(),
        body:    JSON.stringify({ ...input, source_service: input.source_service || "prompt-composer" }),
        signal:  AbortSignal.timeout(TIMEOUT_MS),
      });
      if (!res.ok) {
        // M35.4 — capture raw body for debug, include trace_id
        let detail = "";
        try {
          detail = (await res.text()).slice(0, 500);
        } catch (textErr) {
          detail = `<body read failed: ${(textErr as Error).message}>`;
        }
        logger.warn({ kind: input.kind, status: res.status, detail, trace_id: input.trace_id },
          `audit-gov emit ${input.kind} → ${res.status}`);
      }
    } catch (err) {
      logger.warn({ kind: input.kind, err: (err as Error).message, trace_id: input.trace_id },
        `audit-gov emit ${input.kind} failed`);
    }
  })();
}

File: singularity-iam-service/app/audit_gov_emit.py
"""M22 / fan-out — fire-and-forget emitter to audit-governance-service.

Producers in IAM (auth.* / authz.decision / device-token mint+revoke) push a
single envelope into the central ledger so the platform has one canonical
audit feed. Failure of audit-gov MUST NOT block the request that produced the
event — we log a warn and move on.

This mirrors `mcp-server/src/lib/audit-gov-emit.ts` and
`agent-and-tools/apps/prompt-composer/src/lib/audit-gov-emit.ts` so all three
producers speak the same wire shape.
"""
from __future__ import annotations

import asyncio
import logging
import os
from typing import Any, Optional

import httpx

log = logging.getLogger(__name__)

_AUDIT_GOV_URL = os.environ.get("AUDIT_GOV_URL", "http://host.docker.internal:8500")
_TIMEOUT_SECONDS = 5.0


def emit_audit_event(
    *,
    kind: str,
    source_service: str = "iam",
    trace_id: Optional[str] = None,
    subject_type: Optional[str] = None,
    subject_id: Optional[str] = None,
    actor_id: Optional[str] = None,
    capability_id: Optional[str] = None,
    tenant_id: Optional[str] = None,
    severity: str = "info",
    payload: Optional[dict[str, Any]] = None,
) -> None:
    """Fire-and-forget post into audit-gov.

    Callers must never `await` this; we schedule it on the event loop so it
    completes after the response is returned to the client.
    """
    if not _AUDIT_GOV_URL:
        return
    envelope = {
        "trace_id":       trace_id,
        "source_service": source_service,
        "kind":           kind,
        "subject_type":   subject_type,
        "subject_id":     subject_id,
        "actor_id":       actor_id,
        "capability_id":  capability_id,
        "tenant_id":      tenant_id,
        "severity":       severity,
        "payload":        payload or {},
    }
    try:
        loop = asyncio.get_event_loop()
    except RuntimeError:
        # No running loop (e.g. test sync context) — fall back to a one-shot
        # request synchronously. We still don't raise on failure.
        try:
            with httpx.Client(timeout=_TIMEOUT_SECONDS) as client:
                client.post(f"{_AUDIT_GOV_URL.rstrip('/')}/api/v1/events", json=envelope)
        except Exception as exc:
            log.warning("audit-gov emit %s failed: %s", kind, exc)
        return
    loop.create_task(_emit(envelope))


async def _emit(envelope: dict[str, Any]) -> None:
    try:
        async with httpx.AsyncClient(timeout=_TIMEOUT_SECONDS) as client:
            res = await client.post(
                f"{_AUDIT_GOV_URL.rstrip('/')}/api/v1/events", json=envelope,
            )
            if res.status_code >= 400:
                log.warning("audit-gov emit %s → %s: %s", envelope.get("kind"), res.status_code, res.text[:200])
    except Exception as exc:
        log.warning("audit-gov emit %s failed: %s", envelope.get("kind"), exc)
